Added a Ziggurat normal engine

jkrempus opened the pull request:
I added the NormalZigguratEngineImpl template and an alias for it named NormalZigguratEngine64. I did it that way because my Ziggurat algorithm implementation takes one integer as a template parameter in addition to a floating point type.

I also changed API of the rest a bit. I changed the order of template parameters so that those that can be infered from function parameters come after those that can't be inferred. That way the user doesn't need to specify the types that can be inferred if he uses something other than default for the parameters that can't be inferred. I also changed the order of function parameters for the 4 parameter version of  normal() so that the first three parameters are the same as the parameters of the 3 parameter version. I also removed the default parameter value from the 3 parameter version and added a 2 parameter version. I needed to do that because the default parameter caused compile time errors when the function was instantiated with a RNG type other than Random. 

I also added normalRNG().

jkrempus committed:
Added NormalZigguratEngine

WebDrake commented:
Re template parameters -- I realized that as-is we have an issue where the user passes T as a non-floating point value.  e.g. calling normal(3, 10) will fail because 3, 10 are ints and that overrides any default parameter set.

I'm partly minded to have a RealT template parameter (in addition to T for input type) whose default is real but which can be overriden by the user with another floating point type; and I'm partly minded to take the alternative of saying all these values are just real (as with e.g. mathematical functions).  What do you think?

jkrempus commented:
Maybe we should do it the way it's done in uniform(). Use two template type parameters and have a constraint `isFloatingPoint!(CommonType!(T1, T2))`.

WebDrake commented:
On 11/12/2012 11:35 PM, jerro wrote:

> Maybe we should do it the way it's done in uniform(). Use two template type
> parameters and have a constraint |isFloatingPoint!(CommonType!(T1, T2))|.

Yes, but that still won't deal with the case where they're both integral types, no?

WebDrake commented:
Just to clarify -- the T parameter in my code was never meant to be principally 
the input type, but the return type (and the internal processing type).

jkrempus commented:
Yeah, probably the constraint should be is(typeof(CommonType!(T1, T2))) and the return type and the type used internally would be selected like this:

``` d
alias CommonType!(T1, T2) C;
alias Select!(isFloatingPoint!C, C, real) T;
```

The problem I see with having T as the first parameter and requiring the user to explicitly specify it in some cases is that this is different from the way things work in the rest of Phobos, which makes it unintuitive. I also think it's less convenient than letting the type be inferred from usage,  because most of the time the user will want the return type and the type used internally to be the same as the parameter types, if the parameter types are floating point. If the parameter types are integral, we can just alias T to real like I did in the sample above.

I don't think we should just make the type real as it is done with the math functions. Those functions typically don't have generic implementations - most of them use inline assembly that works with reals internally, so it makes sense for the parameter types to be real, too. normal(), on the other hand, can have a generic implementation, so it seems wrong to force it to work with reals only. I think doing that could probably hurt performance significantly in some cases, as it would require the numbers to be moved between XMM registers and x87 registers if user's code used floats or doubles. Making the floating point type always real would also make some optimizations in the implementation of normal engines impossible - in the Ziggurat engine, for example, I only use one uniform RNG sample to generate one float and one integer if the engine operates on floats. I can do that because the integer I need has less bits than 32 - float.mant_dig. If the engine always operated on reals, I would need at least two uniform RNG samples for that, or maybe even three, if we wanted to have better quality of the normal RNG. 

I'm not sure whether we should use one template parameter for both function parameters or one for each. Both approaches seem to be common in Phobos. uniform() uses two and normal() is similar to that, so maybe we should use two, too. But this can make the template bloat worse (only in cases where the user passes parameters of different types, though). But in any case, I think we should make the API similar to the rest of Phobos, because that is what the users will expect.

WebDrake commented:
Let me lay out a little bit of background here.  My starting point for this API was the proposed C++ standard (linked to from the std.normal documentation) which in turn is pretty closely based on Boost.Random.

The basic formulation for this template is,

``` C++
template <class RealType = double>
class normal_distribution
```

and they don't overtly consider the possibility of multiple different internal engines.  (I wouldn't have either, but that I read the research paper I already sent to you, which indicated the value of different generation schemes in different circumstances.)  So, I took my starting point to be that I should follow this template _plus whatever extras were needed_, in other words that there should be a user-specified type for the core calculation and return value with a default in the case they didn't specify.  I also felt that this default should be the most accurate floating-point type available.

Now, to your points.  First, I agree that the code should be generic and not require the use of real -- I was floating that (ha!) as an idea rather than being wedded to it.  Second, it seems clear to me that we _do_ need a specified default type for the internal calculations and return value.

Here's why.  We basically have 2 possible cases for the input values: either (i) at least one of the passed values is floating-point in which case we can use CommonType!(T1, T2); or (ii) neither of them is floating point, in which case we need a default FP parameter type which can nevertheless be tweaked by the user if they wish.  On balance I feel that the default should be double, which would be the same inferred type as if you called e.g. normal(0.0, 1.0) and is accurate enough for most use cases.

So, the idea would be:
- You pass FP input type(s): normal infers return type from their common type
- You pass integer input types: normal uses default floating-point type (double)
- You specify floating-point type: normal uses your specified type, whatever you pass

I'll follow up tomorrow with a sketch for the updated API, but does this sound like a correct approach?

WebDrake commented:
I've pushed an updated version to my branch: https://github.com/WebDrake/phobos/commit/365619463780aa9fb015d944b62ac4351906be53

This adds some extra checks to all of the normal(...) functions as well as allowing the user to specify a default RealT.  In practice, the type assigned to the engine is determined by, in order of priority:
- Unqual!(CommonType!(RealT, T1, T2)) if RealT is specified and is floating point;
- Unqual!(CommonType!(T1, T2)) if RealT is not specified but CommonType!(T1, T2) is floating point;
- double otherwise.

I've added similar tweaks to the normalRNG function.

Concerns about this version:
- I'm not certain that calling normal!double(0.0, 1.0) and normal(0.0, 1.0) will invoke the same static NormalRandomNumberEngine, whereas in theory at least if the ReturnType is identical, the same instance of the engine should be used.
- The unittests for normalRNG fail with the message: "Warning: statement is not reachable".  I can't see the cause, but maybe you can spot it.
- I tried to prepare two versions of normalRNG with and without the RealT template parameter, in keeping with the first 2 versions of normal(...) (lines 1326-1339), however, the compiler rejected this claiming that when normalRNG() was called without specified template parameters, it was not possible to decide between the versions.  I'm a bit confused why this approach works for normal() but not for normalRNG() ...

Having done this, part of me feels that after all the RealT may be overkill and we could delete it and simply rely on the choice between Unqual!(CommonType!(T1, T2)) or double.  But I'll let you review it and give feedback before reaching a decision.

jkrempus commented:
I think that having RealT, T1 and T2 is too complicated, non-intuitive and inconsistent with the rest of Phobos. I don't think that being able to write `normal!Type(...)` justifies the added complexity. In most cases when common type of parameters is floating point, this is actually the type that the function should operate on, and when they are integral, double is a good default. If the user wants something other than the default, it isn't that hard to convert the parameters to whatever type is desired.

jkrempus commented:
> I'm not certain that calling normal!double(0.0, 1.0) and normal(0.0, 1.0) will invoke the same static NormalRandomNumberEngine, whereas in theory at least if the ReturnType is identical, the same instance of the engine should be used.

This is a good point and we should think about that more. The way normal() is currently written, each instance of normal() function will have its own instance of engine. In other words, every combination of parameter types will have it's own instance, not just every return type. We could work around that by having something like this:

``` d
struct StaticInstance(T)
{
    static T instance;
}

//inside normal():
alias StaticInstance!(NormalRandomNumberEngine!ReturnType).instance engine;

```

jkrempus commented:
> The unittests for normalRNG fail with the message: "Warning: statement is not reachable". I can't see the cause, but maybe you can spot it.

In `assert` on line 1461 there should be `== float`, not `== double`. Because that assert always fails, the statements after that are not reachable.

jkrempus commented:
> I'm a bit confused why this approach works for normal() but not for normalRNG() ...

I don't know know the answer to that, but I have noticed that in the first version both mean and sigma have type T1, so the type T2 can not be inferred from usage and the second version of normal is instantiated in all your unit tests.

WebDrake commented:
> I think that having RealT, T1 and T2 is too complicated, non-intuitive and inconsistent with the rest of Phobos.

Yes, having gone through the implementation, I now agree.  It doesn't add enough to be worth keeping.  I've removed it and kept simply the checks on the common type of T1, T2.

> We could work around that by having something like this

OK, I've added the StaticInstance as you suggest.  I'd like to make sure there is a unittest confirming that each static instance is unique per ReturnType -- any ideas?

> in the first version both mean and sigma have type T1

Ouch!  Nasty typo.  Fixed.

Updated code is now here: https://github.com/WebDrake/phobos/commit/a706785f384075210fe7cf743693232d93bb7ed4

I've also pushed a subsequent patch making StaticInstance private to the module, although I wonder whether there's a case for it being added elsewhere in Phobos as a generally useful tool.  I was wondering the same about some of the supporter functions your Ziggurat implementation includes -- can we think about targeting some of them to better locations in Phobos?

I'd also be inclined to rename randomFloat, randomInt and randomIntAndFloat to something which makes clear they are part of Ziggurat -- or else, if they are more broadly useful functions, to name them more descriptively.  What do you think?

jkrempus commented:
> OK, I've added the StaticInstance as you suggest. I'd like to make sure there is a unittest confirming that each static instance is unique per ReturnType -- any ideas?

If I understand correctly, you'd like to check that there is exactly one distinct static instance of the engine being used for each return type. There is no nice way to do that. You need to get the engine's address out of `normal()` somehow. The only way to do that without modifying `normal` would be to write a dummy engine for that purpose. If you only wanted to test the version of `normal` that takes a RNG as a parameter then I think this would be the simplest way:

``` d
// inside a unittest block:
static struct DummyRng
{
    Xorshift128 rng;
    alias rng this;
    void* p;
}

static struct DummyEngine(T)
{
    T opCall(Rng)(ref Rng rng)
    {
        rng.p = cast(void*) &this;
        return 0;
    }
}

DummyRng rng;
normal!DummyEngine(0, 1, rng);
// rng.p now contains the address of the static instance
```

> I've also pushed a subsequent patch making StaticInstance private to the module, although I wonder whether there's a case for it being added elsewhere in Phobos as a generally useful tool.

This version may be useful elsewhere, but we may want to customize it a bit by adding a static constructor that initializes the engine. I'm not sure if it is a good idea to use a static constructor for initialization, though.

> I was wondering the same about some of the supporter functions your Ziggurat implementation includes -- can we think about targeting some of them to better locations in Phobos?

I think it may be a good idea to have `isPowerOfTwo` as a public, documented part of Phobos, but I don't know in which module it should go.

I don't think `randomInt`, `randomFloat` and `randomIntAndFloat` and others are suitable to be a public part of Phobos, but they may be useful in other distribution engines. I'm not sure what we should do with those.

I don't think any of the other functions are very usefule outside of the Ziggurat engine. It may be a good idea to prefix them with "ziggurat". Another solution would be to put them inside a Ziggurat template (that template would be used as a namespace). Yet another way to solve this would be to create a new file (or directory?) in internal and put entire Ziggurat engine and any other more complex engines there. 

Edit: Wrapping functions in a template does not work because I get errors about not being able to pass local symbols to non global templates.

jkrempus commented:
I noticed we have a bug: `initialized` is still a static variable inside `normal`, so we have multiple instances of `intialized` for each static engine instance. `initialized` should go inside `StaticInstance` or be replaced with a static constructor.

WebDrake commented:
> I think it may be a good idea to have isPowerOfTwo as a public, documented part of Phobos, but I don't know in which module it should go.

I would assume std.traits with most of the other isSomething(...) functions.  Why not request guidance on this when we submit the merge request to Phobos?

> I don't think randomInt, randomFloat and randomIntAndFloat and others are suitable to be a public part of Phobos, but they may be useful in other distribution engines. I'm not sure what we should do with those.

Well, assuming they have some general application, I would name them according to the distribution they are intended to produce.  It's not entirely clear to me what randomInt gives you, for example, apart from a speeded-up uniformly-distributed number, and it's not entirely clear to me that the (r.front - r.min) & (n - 1) formulation is safe in terms of generating a genuinely uniform number (I guess this is what your rngMask and isPowerOfTwo are checking?).  Likewise randomFloat seems to be a (speeded-up?) uniform float in the half-open interval [0, 1).

I don't know whether these speed optimizations really justify avoiding the easy-to-understand (and hopefully well-tested) existing Phobos functions -- and I wonder if it might be better to add those optimizations to the existing uniform distribution rather than introduce new functions here?

But, in the meantime, you might name them something like speedyUniform0toN and speedyUniform01, or whatever.

Or have I misunderstood these functions?  Always possible. :-P

Now, to unittests:

> If I understand correctly, you'd like to check that there is exactly one distinct static instance of the engine being used for each return type. There is no nice way to do that.

What do you think of the following?  It's not "nice" inasmuch as it is quite long, but I think it works.  It relies on the fact that the normal engine (at least Box-Muller) generates and caches a pair of values at a time -- so, you can test if a new instance of the engine is being generated based on whether a call to the function produces the second cached value, or a different number.

And, so far as I can see, it confirms that same type == same instance. :-)

``` D
    {
        auto rng = Random(0);

        auto normalDouble = normalRNG(0.0, 1.0);
        auto d1 = normalDouble(rng);
        auto d2 = normalDouble(rng);
        writeln("d1 = ", d1, ", d2 = ", d2);
        assert(is(typeof(d1) == double));
        assert(is(typeof(d2) == double));

        rng.seed(0);
        rng.popFront();
        rng.popFront();
        auto normalFloat = normalRNG(0.0f, 1.0f);
        auto f1 = normalFloat(rng);
        writeln("f1 = ", f1);
        assert(is(typeof(f1) == float));

        rng.seed(0);
        rng.popFront();
        rng.popFront();
        auto normalReal = normalRNG(0.0L, 1.0L);
        auto r1 = normalReal(rng);
        writeln("r1 = ", r1);
        assert(is(typeof(r1) == real));

        rng.seed(0);
        auto t1 = normal(0.0, 1.0, rng);   // these two calls should use
        auto t2 = normal(0.0f, 1.0, rng);  // the same static engine (double)
        writeln("t1 = ", t1, ", t2 = ", t2);
        assert(is(typeof(t1) == double));
        assert(is(typeof(t2) == double));
        assert(t1 == d1);
        assert(t2 == d2);

        rng.seed(0);
        auto u1 = normal(0.0, 1.0, rng);   // these two calls shoud also use
        auto u2 = normal(0, 1, rng);       // the same static engine (double)
        writeln("u1 = ", u1, ", u2 = ", u2);
        assert(is(typeof(u1) == double));
        assert(is(typeof(u2) == double));
        assert(u1 == d1);
        //assert(u1 == d2);

        rng.seed(0);
        auto v1 = normal(0.0, 1.0, rng);   // should use double engine
        auto v2 = normal(0.0f, 1.0f, rng); // should use new (float) engine
        auto v3 = normal(0.0, 1.0, rng);   // should use double engine
        writeln("v1 = ", v1, ", v2 = ", v2, ", v3 = ", v3);
        assert(is(typeof(v1) == double));
        assert(is(typeof(v2) == float));
        assert(v1 == d1);
        assert(v2 != d2);
        assert(v2 == f1);
        assert(v3 == d2);

        rng.seed(0);
        auto w1 = normal(0.0, 1.0, rng);   // should use double engine
        auto w2 = normal(0.0, 1.0L, rng);  // should use new (real) engine
        auto w3 = normal(0.0, 1.0, rng);   // should use double engine
        writeln("w1 = ", w1, ", w2 = ", w2, ", w3 = ", w3);
        assert(is(typeof(w1) == double));
        assert(is(typeof(w2) == real));
        assert(w1 == d1);
        assert(w2 != d2);
        assert(w2 != v2);   // because real is higher-precision than float
        assert(w2 == r1);
        assert(w3 == d2);
    }
```

Note that the writeln() statements in there are just for test/visualization purposes and will be removed in any patch to the unittests.

So far this has just been tested with BoxMuller -- how about making it a template/template mixin so that we can test it for both Box-Muller and Ziggurat, and for any other engines based on generating a pair of random variates at a time?  I'm not sure how the above might be affected by the initialize() bug you noted, which 

On that note:

> I noticed we have a bug: initialized is still a static variable inside normal, so we have multiple instances of intialized for each static engine instance. initialized should go inside StaticInstance or be replaced with a static constructor. 

Why don't you push a patch for this -- I feel you have a better chance to get it right than me :-)

Meanwhile, I'll try and get on with some documentation.

jkrempus commented:
>  would assume std.traits with most of the other isSomething(...) functions

isSomething things in std.traits are all templates that test properties of types. I think it would be better to put isPowerOfTwo in math.d, but maybe there's some other better place. 

> Well, assuming they have some general application, I would name them according to the distribution they are intended to produce. It's not entirely clear to me what randomInt gives you, for example, apart from a speeded-up uniformly-distributed number, 

That's right, performance is the only reason I use those functions. 

> and it's not entirely clear to me that the (r.front - r.min) & (n - 1) formulation is safe in terms of generating a genuinely uniform number (I guess this is what your rngMask and isPowerOfTwo are checking?). 

It is safe. If `rngMask!r` compiles, that means that the number of numbers that the RNG can return (let's call it `m`) is a power of two. I also check that `n` is a power of two, and `m` is not smaller than `n`. Consider a number i, `0 <= i < n`. There are exactly `2 ^^ (log2(m) - log2(n))` numbers `j < m` such that `j & (n - 1) == i`. That's true because the condition  `j & (n - 1) == i` constrains the lowest `log2(n)` bits to a specific value and `j` is a `log2(m)` bit number so the condition leaves `log2(m) - log2(n)` free bits. So if all numbers `j` are equally likely, so are all the numbers `i = j & (n - 1)`.

> I don't know whether these speed optimizations really justify avoiding the easy-to-understand (and hopefully well-tested) existing Phobos functions 

Those optimizations give a rather large performance boost. When using double precision and Mt19937 RNG it takes my current implementation 2.4 s to generate 100000000 samples. If I replace the call to `randomIntAndFloat` (that uses `randomInt` and `randomFloat` internally in this case) with two calls to uniform, the time goes up to 4 s. I get a similar difference in performance when using DMD. The difference is even greater when an additional optimization in randomIntAndFloat can be used - when operating on single precision numbers, the time is 1.8 s for my current implementation and 4 s for one that uses uniform(). The additional optimization in randomIntAndFloat will apply to doubles, too, if someone writes a 64 bit RNG engine. 

> and I wonder if it might be better to add those optimizations to the existing uniform distribution rather than introduce new functions here?

That can't be done. `randomInt` relies on the fact that `n` is known at compile time and `randomFloat` always returns numbers between 0 and 1.

> But, in the meantime, you might name them something like speedyUniform0toN and speedyUniform01, or whatever.

I have renamed them to fastUniformInt, fastUniformFloat and fastUniformIntAndFloat now.

> So far this has just been tested with BoxMuller -- how about making it a template/template mixin so that we can test it for both Box-Muller and Ziggurat, and for any other engines based on generating a pair of random variates at a time?

Ziggurat engine doesn't actually cache any numbers. As for the rest, I don't think we need to put much effort into testing this. The language guarantees that there will be exactly one instance of template `StaticInstance` for each template parameter and it also guarantees there will be one static field `instance` for each instance of template `StaticInstance`. Testing something even though you know it should work is fine, but I really don't think we should be so concerned about this. Testing it with just one engine is enough, IMO.

> Why don't you push a patch for this -- I feel you have a better chance to get it right than me :-)

I opened a pull request that fixes this and some other stuff related to Ziggurat engine. I fixed it by moving the initialization into a static constructor. This is an implementation detail and can easily be changed later if need be.

WebDrake commented:
> It is safe. If |rngMask!r| compiles, that means that the number of numbers that
> the RNG can return (let's call it |m|) is a power of two. I also check that |n|
> is a power of two, and |m| is not smaller than |n|. Consider a number i, |0 <= i
> < n|. There are exactly |2 ^^ (log2(m) - log2(n))| numbers |j < m| such that |j
> & (n - 1) == i|. That's true because the condition |j & (n - 1) == i| constrains
> the lowest |log2(n)| bits to a specific value and |j| is a |log2(m)| bit number
> so the condition leaves |log2(m) - log2(n)| free bits. So if all numbers |j| are
> equally likely, so are all the numbers |i = j & (n - 1)|.

Cool.  Not meaning to sound like I was doubting you -- just wanting to make sure 
we have the justification clear and in place in advance of submitting this to 
Phobos.

> Those optimizations give a rather large performance boost.

Excellent.

> I have renamed them to fastUniformInt, fastUniformFloat and
> fastUniformIntAndFloat now.

Sounds good to me.

> Ziggurat engine doesn't actually cache any numbers. As for the rest, I don't
> think we need to put much effort into testing this. The language guarantees that
> there will be exactly one instance of template |StaticInstance| for each
> template parameter and it also guarantees there will be one static field
> |instance| for each instance of template |StaticInstance|. Testing something
> even though you know it should work is fine, but I really don't think we should
> be so concerned about this. Testing it with just one engine is enough, IMO.

Yes, I know it's kind of overkill.  It was just fun to work out the test to make 
sure. :-)

> I opened a pull request that fixes this and some other stuff related to Ziggurat
> engine. I fixed it by moving the initialization into a static constructor. This
> is an implementation detail and can be changed later if need be.

That's great.  As you'll see I have one small query over that -- apart from 
this, I think that all we really need now is documentation!

jkrempus commented:
By the way, I have implemented one of the tests in the paper you have linked to, so you can use that for testing distributions. The code is [here](https://github.com/jerro/test-distributions/blob/master/test.d). It depends on [dstats](git://github.com/dsimcha/dstats.git). Unfortunately there's a bug in gammaIncompleteCompl in Phobos, so you'll also need to download [this file](https://github.com/jerro/phobos/blob/upstream-master/std/internal/math/gammafunction.d) and put it on the command line to make the test work. 

Edit: There's a problem with this test that I need to fix.
Edit 2: It's fixed now.

WebDrake commented:
Nice -- I will take a look at this probably next week :-)  Do you have any results out of it yet regarding your Ziggurat and my (well, Boost's...) Box-Muller?

jkrempus commented:
I used a goodness of fit test described at pages 24 and 25 of the paper titled "Gaussian Random Number Generators
". For single precision, Ziggurat engine failed at `2 ^^ 33` samples and Box Muller passed the tests for up to `2 ^^ 35`. I didn't run the test of Box Muller engine with a high enough number of samples to see it fail, because at `2 ^^ 35` samples, the test already took about 40 minutes. 

I had a problem with this test that I have since fixed, but I think knowing the reasons behind that problem can help in the interpretation of the results.  The problem was that all tests were failing at `2 ^^ 30` or `2 ^^ 31` samples. I knew that something had to be wrong with the tests, because that happened even for uniform distribution, for both Xorshift and Mersenne Twister RNG. 

The reason behind this turned out to be that I was putting the samples in bins that were supposed to have equal probability. Then I used chi suared test to check whether the samples were equally distributed. This would have worked fine if floating point numbers were continuous, but they aren't. Because the bin boundaries that would have given equal probability bins could not be represented as a floating point number, the actual bin probabilities were not equal. When I used the actual single precision bin boundaries to calculate bin probabilities, the problem went away. 

The problems I had with the tests at over `2 ^^ 30` samples show that those tests measure resolution and accuracy at a level where the assumption that single precision numbers approximate real numbers well enough already breaks. I had to take the fact that floating point numbers can only take discrete values into account to be able to execute those tests. If you tried to use single precision normal random generators for Monte Carlo integration and wanted to get similar resolution and precision, you would need do the same in all calculations you would do on the samples. I think that at that point, you would be _far_ better of with double precision numbers. So I think that for single precision, if a generator passes this test at over `2 ^^ 30` samples, that's good enough for any real world use.

For double precision floating point numbers, I haven't been able to get the test to fail for any distribution. The most I have tried was `2 ^^ 37` samples with Ziggurat engine and Xorshift128 generator. That took about two hours and passed. All the RNGs in Phobos currently produce 32 bit numbers AFAIK. If we used 64 bit RNGs, we could probably get far better quality of the distribution than we can with 32 bit RNGs. But if we can't run tests with high enough numbers of samples to get the normal RNGs that use 32 bit unfiorm RNGs to fail, we can't really test that assumption.

WebDrake commented:
Interesting results.  It's interesting that you say that Box-Muller passes for 2^^36 or more samples at double precision; I'm not sure that I understand what they mean by "standard" versus "full-fraction" integer-to-FP conversion, but I guess your result could reflect the numerical precision of the Phobos uniform RNG.

So, I'm curious to see if (just for an experiment) you replace your fast RNG functions with Phobos' uniform, whether that affects Ziggurat's performance relative to Box-Muller.  We'd _expect_ Ziggurat to be more accurate, after all.  (As an alternative experiment, you could replace Box-Muller's calls to uniform(0, 1) with calls to your fastUniformFloat, and see whether that LOWERS Box-Muller's statistical performance.)

You could also try undoing your speed tweaks to Ziggurat -- the layer size choices that we discussed -- just to see if that makes a difference to Ziggurat's performance.

That said, my suspicion is that probably single-precision FP is in any case too low-res to be valid for use in large-scale statistical work.

BTW I think that picking up on the limits of floating-point precision is probably kind of important, so I'm not sure I'd want to tweak the bins as you did -- did these issues also extend to the double-precision tests?

jkrempus commented:
> I'm not sure that I understand what they mean by "standard" versus "full-fraction" integer-to-FP conversion, but I guess your result could reflect the numerical precision of the Phobos uniform RNG.

They explained that on pages 18 and 19. Basically standard conversion means converting the integer sample to floating point and dividing it by RNG's range. If you are converting a 32 bit integer to single precision floating point, this will give you full resolution of floating point near 1, but not near zero. Consider, for example the interval between 0 and `2 ^^ (-20)`. The resolution of single precision floating point on that interval is at least `2 ^^ (-44)`, but the resolution of numbers calculated using "standard" conversion is only `2 ^^ (-32)`, so standard conversion doesn't give you full resolution of single precision floating point o nthat interval. Full fraction conversion is supposed to give you full resolution. One way to do that for single precision numbers would be to generate random 23 bits for mantissa and then generate a random exponent so that the probability for each value of the exponent is proportional to the range of numbers that exponent covers. 

> So, I'm curious to see if (just for an experiment) you replace your fast RNG functions with Phobos' uniform, whether that affects Ziggurat's performance relative to Box-Muller

I have tried that and it doesn't make a difference. 

>  We'd expect Ziggurat to be more accurate, after all

I'll try to investigate that.

> (As an alternative experiment, you could replace Box-Muller's calls to uniform(0, 1) with calls to your fastUniformFloat, and see whether that LOWERS Box-Muller's statistical performance.)

I haven't tried that yet, but I will.

> You could also try undoing your speed tweaks to Ziggurat -- the layer size choices that we discussed -- just to see if that makes a difference to Ziggurat's performance.

That requires a lot of changes, but maybe I'll do it anyway. 

> That said, my suspicion is that probably single-precision FP is in any case too low-res to be valid for use in large-scale statistical work.

I'm sure there are many cases where single precision is good enough, but I do think that double precision should be the default choice. Single precision is obviously not usable when you need high resolution and I think that the problem I had with the test demonstrates that nicely.

> BTW I think that picking up on the limits of floating-point precision is probably kind of important, so I'm not sure I'd want to tweak the bins as you did

If I didn't do that, I couldn't test anything past `2 ^ 30` samples since the test would fail no matter how good the generator was. It would only work in rare cases when the bin boundaries could be represented exactly, for example when testing uniform distribution with a number of bins that is a power of two. 

As for the limits of floating point precision, I think that can be estimated in this case. Basically, if you are making a histogram and relative error in bin size due to finite resolution of floating point is not small in comparison to relative error in measured probability due to randomness, then you have reached the limits of floating point precision. In my case, that happens around `2 ^^ 30` samples. The probability for each bin at that size is about `1 ^^ (-18)` and the value of normal function is about 1 / 4, so the bin width is around `2 ^^ (-16)`. The resolution of single precision floating point at 1 is `2 ^^ (-24)`, so relative error of bin width due to limited resolution is `2 ^^ (-8)`. Expected number of samples per bin is `2 ^^ 12`, so relative error in probability due to randomness is `2 ^^ (-6)`. `2 ^^ (-8)` not insignificant when compared  to `2 ^^ (-6)` - it is true that we couldn't notice the additional error just by looking at one bin, but because there are many bins, chi squared test can catch it.

> did these issues also extend to the double-precision tests

I noticed no such problems with double precision. I wouldn't expect to, either. If I repeat the calculation from the previous paragraph for double precision and `2 ^^ 40`, I get relative error due to limited resolution `2 ^^ (-31)` and relative error due to randomness `2 ^^ (-8)`. So even if I ran a test with `2 ^^ 40` samples I would still be far from having problems with resolution of double precision floating point and running such test would take a whole day on my machine.

WebDrake commented:
On 11/22/2012 09:11 PM, jerro wrote:

> > You could also try undoing your speed tweaks to Ziggurat -- the layer size
> > choices that we discussed -- just to see if that makes a difference to
> > Ziggurat's performance.
> 
> That requires a lot of changes, but maybe I'll do it anyway.

Well, I fear it's probably necessary, not as a block to getting this stuff 
submitted to Phobos but just as a check to ensure that your changes didn't cause 
a dip in accuracy.  After all, the Ziggurat method has a particular standard 
meaning and we don't want D to give people the impression that it's equivalent 
to the standard algorithm if that's not the case.

> > BTW I think that picking up on the limits of floating-point precision is
> > probably kind of important, so I'm not sure I'd want to tweak the bins as
> > you did
> 
> If I didn't do that, I couldn't test anything past |2 ^ 30| samples since the
> test would fail no matter how good the generator was. It would only work in rare
> cases when the bin boundaries could be represented exactly, for example when
> testing uniform distribution with a number of bins that is a power of two.

Well, that's what I mean -- that the precision of the floating-point means the 
quality of pseudo-random number generation hits a strict upper bound before the 
generator's own limits are reached.

> I noticed no such problems with double precision. I wouldn't expect to, either.
> If I repeat the calculation from the previous paragraph for double precision and
> |2 ^^ 40|, I get relative error due to limited resolution |2 ^^ (-31)| and
> relative error due to randomness |2 ^^ (-8)|. So even if I ran a test with |2 ^^
> 40| samples I would still be far from having problems with resolution of double
> precision floating point and running such test would take a whole day on my
> machine.

Excellent.

I'm still curious as to why you didn't see any error with Box-Muller at the 
numbers of samples cited in Table IV, but I think your results give us reason to 
feel that we have a couple of good-quality algorithms here.

jkrempus commented:
> Well, I fear it's probably necessary, not as a block to getting this stuff
> submitted to Phobos but just as a check to ensure that your changes didn't cause
> a dip in accuracy.  After all, the Ziggurat method has a particular standard
> meaning and we don't want D to give people the impression that it's equivalent
> to the standard algorithm if that's not the case.

I have [a branch that uses a basic Ziggurat algorithm](https://github.com/jerro/phobos/blob/standard-ziggurat/std/random.d#L1651) now. It isn't very optimized and there is some cruft left in from the modified algorithm that isn't used in the standard one. But the code that is executed does perform a standard Ziggurat algorithm. The qulity isn't any better than that of the modified algorithm.

I have noticed that the quality depends on the number of layers. With 2 layers, standard Ziggurat fails at `2 ^^ 30` samples and with 4096 layers, it fails at `2 ^^ 35`.

WebDrake commented:
Sorry for the long delay in getting back to you on this -- work, plus the lovely viruses of the season ... ;-)  Anyway: I've pushed an update with basic documentation for normal random number generation.

WebDrake commented:
Further note -- if I understand correctly, your fastUniformFloat is almost identical to Boost's uniform_01 distribution.  Compare your function to this (copy-pasted from Boost 1.49):

``` C++
  template<class Engine>
  result_type operator()(Engine& eng) {
    for (;;) {
      typedef typename Engine::result_type base_result;
      result_type factor = result_type(1) /
              (result_type((eng.max)()-(eng.min)()) +
               result_type(std::numeric_limits<base_result>::is_integer ? 1 : 0));
      result_type result = result_type(eng() - (eng.min)()) * factor;
      if (result < result_type(1))
        return result;
    }
  }
```

So there may really be a value in renaming it to uniform01 for consistency.  I don't know whether you read anything into their choice of using the +1 denominator alteration only in the case of the RNG's base type being an integer -- a tweak worth making?

Note also the comment on Boost's version: "The current implementation is buggy, because it may not fill all of the mantissa with random bits. I'm unsure how to fill a (to-be-invented) boost::bigfloat class with random bits efficiently. It's probably time for a traits class."

jkrempus commented:
> So there may really be a value in renaming it to uniform01 for consistency.

I'm don't see a reason why consistency with Boost would be important to us. But if you want,  I can rename it to uniform_01.

> I don't know whether you read anything into their choice of using the +1 denominator alteration only in the case of the RNG's base type being an integer -- a tweak worth making?

I already have +1 in the denominator like they do. 

> Note also the comment on Boost's version: "The current implementation is buggy, because it may not fill all of the mantissa with random bits. I'm unsure how to fill a (to-be-invented) boost::bigfloat class with random bits efficiently. It's probably time for a traits class."

This is about the difference between full-fraction and standard conversion that we have already talked about. It is something to keep in mind, but I think that if we ever add uniform random number generators with fully random fraction, their use in other distributions should be optional as they will probably perfrom a lot worse and the difference in resolution will not be important for most users. 

Besides, if we care so much about resolution of random number distributions, we need to improve the resolution of unform for double and real types first. The result of the current implementation never has more random bits than log2(rng.max - rng.min) in the mantissa. Because real has 64 mantissa bits, double has 53, and all the random number engine in std.random are 32 bit,  the result of uniform for double and real arguments has much worse resolution than what the floating point type supports. 

By the way, I found out what is the reason that the distribution quality tests fail at about 2^^34 samples for the ziggurat engine. It is, again, caused by the limits of  single precision arithmetic and it has to do with scaling. 

Imagine you have a generator that returns floating point numbers between 0 and 1 and the probability for each floating point number is proportional to the size of the interval that it covers. Now let's say you multiply the results with some number that doesn't have nice binary representation, for example, let's say you multiply them with 0.9. Consider what the multiplication does to numbers between 0.6 and 1.0 (I have chosen this interval because the resolution of floating point numbers is constant on this interval and also on the interval betwen 0.54 and 0.9). The multiplication will map some of the numbers to distinct numbers, but some pairs of adjecent floating point numbers are mapped to the same   number. So if all the numbers between 0.6 and 1.0 were equally probable, some of the results of multiplication will be twice as probable as others. 

This problem does not also affect Ziggurat engine, it affects any code that multiplies random numbers with numbers with anything that isn't a power of two. I have tested uniform() with interval [0, 0.9), and it failed somewhere around 2^^30 samples IIRC. Ziggurat engine does a bit better than uniform(), and the more layers it uses the better it does. The reason for this is that in different layers, different numbers will have higher probability, so the distribution partially evens out.

I made [a branch](https://github.com/jerro/phobos/blob/normal-accurate-uniform/std/random.d) where Ziggurat engine uses `accurateUniformFloat`, which doesn't doesn't have the problem I described above. It does much better on my tests, but is also much slower.

WebDrake commented:
On 12/18/2012 10:46 PM, jerro wrote:

> I'm don't see a reason why consistency with Boost would be important to us. But
> if you want, I can rename it to uniform_01.

Let's leave this as a question at pull-request time, then.  Personally I think 
that naming consistency (or similarity) is nice to have when it doesn't cost you 
anything and when it highlights points of similarity to other codebases.

> I already have +1 in the denominator like they do.

I know, I'm thinking about dealing in advance with the case where someone 
implements an RNG whose base type is _not_ an integer (e.g. the lagged_fibonacci 
generators in Boost).

> Besides, if we care so much about resolution of random number distributions, we
> need to improve the resolution of unform for double and real types first. The
> result of the current implementation never has more random bits than
> log2(rng.max - rng.min) in the mantissa. Because real has 64 mantissa bits,
> double has 53, and all the random number engine in std.random are 32 bit, the
> result of uniform for double and real arguments has much worse resolution than
> what the floating point type supports.

Fair enough.  I think this (and your results on distribution quality for both 
normal and uniform distributions) may be worth raising in discussion on the D lists.

> By the way, I found out what is the reason that the distribution quality tests
> fail at about 2^^34 samples for the ziggurat engine. It is, again, caused by the
> limits of single precision arithmetic and it has to do with scaling.

So from this it follows that we're getting comparable results to the paper -- 
i.e. for single precision Box-Muller fails at 2 ^^ 35 samples (as with the 
full-fraction results of Thomas et al.) and Ziggurat with the accurate uniform 
float passes for more than 2 ^^ 36?

jkrempus commented:
> So from this it follows that we're getting comparable results to the paper -- 
> i.e. for single precision Box-Muller fails at 2 ^^ 35 samples (as with the
> full-fraction results of Thomas et al.) and Ziggurat with the accurate uniform
> float passes for more than 2 ^^ 36?

Not really. Ziggurat engine still fails at 2 ^^ 36 with 128 layers (but passes at 2 ^^ 35 - the branch that doesn't use `accurateUniformFloat` fails at 33). It may still be possible to improve my implementation so that it produces a distribution of better quality, but unless we are fix the scaling issue (which is obviously the main cause of imprecision) it won't make a difference. And I don't think we should fix the scaling issue, because I don't see any practical value in doing that. Even if we fix the issue in the normal engine, the issue will be back as soon as you multiply its output with something. Which means that the output of `normal()` will  suffer from this issue when sigma is not a power of two no matter what we do in the normal engine. The performance cost of fixing this is also very high,  `NormalZigguratEngine` that uses `accurateUniformFloat` is about 4 times slower because of that. 

Box-Muller has passed  for all sizes I have tried it on, which is at least up to 2^^35 (I don't remember what was the largest size, I'm running it now with 2^^37), even though  it uses "standard conversion". Box-Muller in the paper failed at 2^^27 when using "standard conversion". Maybe they are doing something wrong in their Box-Muller implementation? Or is it because you are using higher precision in some places inside your engine ( typeof(sin(float.init)) is real).

WebDrake commented:
> Box-Muller has passed for all sizes I have tried it on, which is at least up to 2^^35 (I don't remember what was the largest size, I'm running it now with 2^^37), even though it uses "standard conversion". Box-Muller in the paper failed at 2^^27 when using "standard conversion". Maybe they are doing something wrong in their Box-Muller implementation? Or is it because you are using higher precision in some places inside your engine ( typeof(sin(float.init)) is real).

Possibly -- I don't feel that I have the expertise to give a definitive answer here.  It's one reason why I think at some point we should reach out to serious stats people for a careful review of Phobos' random number capabilities.

The bottom line here is that I think this code is "good enough" for practical use especially when one considers that most serious applications will be using double or real floating-point types.  So, we should finish documentation and get a pull request submitted, perhaps together with a brief report of your findings here.

Obviously it's nice if we can get Phobos to support the highest quality random number generation, but it's also likely that the most serious statistical use-cases might be better served by a specialist scientific library.  One thing I was hoping to achieve with the idea of internal generation engines was to make it easy to incorporate higher-quality or specialist methods into the existing Phobos framework.

Re your remarks on speed, it might be a worthwhile tweak to replace the calls to uniform() in BoxMuller with calls to fastUniformFloat.

Would you like to draft some documentation and comments for your code?  I can also have a go at this if you're constrained for time.

jkrempus commented:
> Possibly -- I don't feel that I have the expertise to give a definitive answer here. It's one reason why I think at some point we should reach out to serious stats people for a careful review of Phobos' random number capabilities.

Of course, if we can get a statistics expert to take a look at it, we should.

> The bottom line here is that I think this code is "good enough" for practical use especially when one considers that most serious applications will be using double or real floating-point types. So, we should finish documentation and get a pull request submitted, perhaps together with a brief report of your findings here.

I agree.

> Obviously it's nice if we can get Phobos to support the highest quality random number generation, but it's also likely that the most serious statistical use-cases might be better served by a specialist scientific library.

I think a lot could also be added to phobos. For example, `uniform` for floating point types should probably be fixed so that it would use at least as many random bits as there are bits in the mantissa. But this should obviously be done in a separate pull request. We could also add additional optional template parameter to `uniform` so that it could optionally support generating floating point numbers with fully random fraction.

> Re your remarks on speed, it might be a worthwhile tweak to replace the calls to uniform() in BoxMuller with calls to fastUniformFloat.

I have tried replacing the calls to `uniform` with `fastUniformFloat` and the speed went from 13.3 MSamples/s to 15.5. `fastUniformFloat` should not result in worse distribution than `uniform`, but I tested `NormalBoxMullerEngine` using `fastUniformFloat` anyway, with 2^^35 samples, and it passed. So I agree that we should do this.

> Would you like to draft some documentation and comments for your code? I can also have a go at this if you're constrained for time.

I'll do this tomorrow. I'd also like to merge a couple of changes to the code. One is a small optimization in Ziggurat algorithm, and the other one is a change to struct `Normal`. I noticed that removing the enforcement in `normal()` speeds up Ziggurat algorithm by about 50%. The thing is, that enforcement does not need to be done for every call in the struct API, since sigma is only set once in the constructor. So I have added the enforcement to the constructor and changed `opCall` so that it does not call `normal()` any more.

jkrempus commented:
By the way, I have just noticed that single precision Box-Muller engine fails the test at 2^^30 samples when compiled with DMD with -O -inline (I have been using gdc for my previous tests). It succeds at 2^^29 samples, though, so I think It's still good enough, since any single precision engine will fail the test at around 2^^30 samples when sigma is not a power of two (I run most of the tests with mean = 0, sigma = 1).
